bokehjs/src/lib/core/util/spatial.ts

```diff
@@ -51,18 +43,59 @@ class _FlatBush extends FlatBush {
           continue
         }
 
+        const index = this._indices[pos >> 2] | 0
+
         if (nodeIndex < this.numItems * 4) {
-          results.set(index) // leaf item
+          nodeFunction(index, {x0: nodeMinX, y0: nodeMinY, x1: nodeMaxX, y1: nodeMaxY})
         } else {
           queue.push(index) // node; add it to the search queue
         }
+
       }
 
       nodeIndex = queue.pop()
     }
+  }
 
-    return results
+  search_indices(minX: number, minY: number, maxX: number, maxY: number): Indices {
+    const result = new Indices(this.numItems)
+
+    function fn(result: Indices) {
+      return (index: number, _node: Rect) => { result.set(index) }
+    }
+    this.search_apply(minX, minY, maxX, maxY, fn(result))
+
+    return result
   }
+
+  search_bounds(minX: number, minY: number, maxX: number, maxY: number): Rect {
+    const result = empty()
+
+    function fn(result: Rect) {
+      return (_index: number, node: Rect) => {
+        if (node.x0 >= minX && node.x0 < result.x0) {
+          result.x0 = node.x0
+        }
+        if (node.x1 <= maxX && node.x1 > result.x1) {
+          result.x1 = node.x1
+        }
+        if (node.y0 >= minY && node.y0 < result.y0) {
+          result.y0 = node.y0
+        }
+        if (node.y1 <= maxY && node.y1 > result.y1) {
+          result.y1 = node.y1
+        }
+      }
+    }
+    this.search_apply(minX, minY, maxX, maxY, fn(result))
+
+    if (!isFinite(result.x0 + result.y0 + result.x1 + result.y1)) {
+      return empty()
+    }
```

Comment:
I'm not 100% confident that this condition should be needed. Some `x_range` inputs that did not intersect results seemed to come back with e.g. `x0: Infinity` which I think could only happen if some internal boxes can have `Infinity` values. Is that the case?